test/CMakeLists.txt

```diff
@@ -163,15 +163,40 @@ add_test(
 # tests are compiled without compiler extensions to ensure the stlab headers
 # are not dependent upon any such extension.
 #
-set_target_properties(
-  stlab.test.channel
-  stlab.test.future
-  stlab.test.serial_queue
-  stlab.test.cow
-  stlab.test.task
-  stlab.test.tuple
-  stlab.test.traits
-  PROPERTIES CXX_EXTENSIONS OFF )
+
+if (EMSCRIPTEN)
+  set_target_properties(
+    stlab.test.channel
+    stlab.test.future
+    stlab.test.serial_queue
+    stlab.test.cow
+    stlab.test.task
+    stlab.test.tuple
+    stlab.test.traits
+    stlab.test.forest
+    PROPERTIES
+    CXX_EXTENSIONS OFF
+    SUFFIX ".html"
+    COMPILE_OPTIONS "-pthread;-fwasm-exceptions"
+    LINK_OPTIONS "--emrun;-pthread;-fwasm-exceptions;-sINITIAL_MEMORY=300MB;-sPTHREAD_POOL_SIZE=100;-sPROXY_TO_PTHREAD"
```

Comment:
Could you explain why all these things are needed? Also, if it is needed, I think it belongs in a separate toolchain file instead of here.

Reply:
Sure. I'll also make the move to a toolchain file.

### `SUFFIX .html`
Required to configure `emcc` to produce bundles that can be run under `emrun`. It adds hooks to enable capturing stdout and stderr.

### `-fwasm-exceptions`. 
Without this, the tests error out with the following:
```
Pthread 0x005141d0 sent an error! http://localhost:6931/stlab.test.task.worker.js:199: uncaught exception: 10570976 - Exception catching is disabled, this exception cannot be caught. Compile with -sNO_DISABLE_EXCEPTION_CATCHING or -sEXCEPTION_CATCHING_ALLOWED=[..] to catch.
```

### `-sINITIAL_MEMORY=300MB`
Without this, the tests OOM. The first sign is an error out with the following:
```
Pthread 0x0058faf8 sent an error! http://localhost:6931/stlab.test.future.wasm:7450574: RuntimeError: unreachable executed
```
If we run the problematic test under a "head-ful" `emrun`, we can see some javascript errors that explain:
```
Aborted(Cannot enlarge memory arrays to size 17457152 bytes (OOM). Either (1) compile with -
sINITIAL_MEMORY=X with X higher than the current value 16777216, (2) compile with -
sALLOW_MEMORY_GROWTH which allows increasing the size at runtime, or (3) if you want malloc to return 
NULL (0) instead of this abort, compile with -sABORTING_MALLOC=0)
```

I tried smaller values as well during development, but got intermittent failures. Just now, I went back and tried 150MB and 200MB. The former would fail intermittently, but the latter seems stable. I can lower the value if you like. Though, I think having a bit of headroom in the tests will be nice to avoid new OOMs when tests are added.

The fact we have to configure this at all is because pthread support does not mix with `-sALLOW_MEMORY_GROWTH`. See [WASM design issue 1271](https://github.com/WebAssembly/design/issues/1271).

### `-sPTHREAD_POOL_SIZE=100`
Without this, the tests fail to make forward progress, presumably because some tests block the main thread. In revisiting this, though, I found I was able to shrink the pool down considerably to 16 threads. I tried 8 but we did deadlock.

As above, I'm a bit concerned about reliability of these settings as we add more tests. I'd like to give us a bit of headroom so I think I'll go with 32 threads unless you disagree.

### `-sPROXY_TO_PTHREAD`
Without this, we exhaust the thread pool very quickly. The error looks like this:
```
Tried to spawn a new thread, but the thread pool is exhausted.
This might result in a deadlock unless some threads eventually exit or the code explicitly breaks out to the event loop.
```

This setting will wrap our main function inside a pthread, which seems to alleviate the deadlocking. You can read more about the setting [here](https://emscripten.org/docs/porting/pthreads.html#blocking-on-the-main-browser-thread)

Reply:
These explanations are great and should be inline with the code so future readers can understand the rationale.

Reply:
Added to the new toolchain file.